Update LDAPAuthentication to Spring LDAP in order to remove insecure JNDI usage

JohnnyMendesC opened the pull request:
## References
* Fixes #10823

## Description
This PR addresses the potential security vulnerability related to insecure JNDI usage (specifically potential deserialization attacks via user input). It replaces the raw `javax.naming` JNDI implementation in the LDAP authentication module with the secure and well-maintained **Spring LDAP** library. Additionally, it documents why other JNDI usages in the codebase are considered safe.

## Instructions for Reviewers

**Changes in this PR:**

* **Dependency Management:**
    * Upgraded `spring-ldap-core` to version **3.3.5** (in parent `pom.xml` and `dspace-api/pom.xml`) to ensure full compatibility with Spring Boot 3.5.x and resolve classpath conflicts (`NoClassDefFoundError`).
    * Added explicit `micrometer` dependencies to the parent POM to resolve dependency convergence issues.
    * Updated `LICENSES_THIRD_PARTY` to include the Apache 2.0 license for Spring LDAP.

* **LDAP Refactoring (`LDAPAuthentication.java`):**
    * Refactored the `SpeakerToLDAP` inner class to use `LdapTemplate` and `LdapContextSource`.
    * Implemented `EqualsFilter` and `LdapNameBuilder` to ensure user input (`netid`) is properly escaped and DNs are constructed safely, preventing JNDI injection.
    * Replaced manual `InitialLdapContext` handling with Spring's connection pooling and context management.
    * Maintained existing logic for user search, DN resolution, and password verification (bind).
    * Added fail-fast logic: throws `IllegalStateException` if the LDAP provider URL is missing during setup.

* **Test Configuration:**
    * Disabled the LDAP Health Check (`management.health.ldap.enabled=false`) in `application-test.properties` for integration tests. This prevents false negatives (status `DOWN`) in `HealthIndicatorsIT` since a real LDAP server is not present in the test environment.

* **Documentation ("False Positives"):**
    * Added code comments to `DSpaceKernelInitializer.java` and `EmailServiceImpl.java` explaining that `@SuppressWarnings("BanJNDI")` is justified because those lookups rely on internal DSpace configuration (administrator-controlled), not user input.

**Testing & Verification:**

1.  **Code Review:** The primary verification is ensuring that `javax.naming` is no longer used directly for user-supplied data in `LDAPAuthentication`.
2.  **Compilation:** Run `mvn clean package` to ensure the new Spring LDAP dependency is correctly resolved.
3.  **CI/CD:** Verify that Integration Tests (specifically `HealthIndicatorsIT` and `CurationScriptIT`) pass successfully, confirming that the Spring Context loads correctly with the new dependencies.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome).
However, reviewers may request that you complete any actions in this list if you have not done so. If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [ ] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [ ] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [ ] My PR **passes Checkstyle** validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [ ] My PR **includes Javadoc** for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [ ] My PR **passes all tests and includes new/updated Unit or Integration Tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [ ] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [ ] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [ ] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

JohnnyMendesC committed:
Dep: Add Spring LDAP 3.2.15 dependency to replace JNDI

JohnnyMendesC committed:
fix(#10823): Replace JNDI with Spring LDAP in LDAPAuthentication

JohnnyMendesC committed:
Docs: Add comments explaining why BanJNDI is suppressed in safe files

JohnnyMendesC committed:
Fix: Add Spring exclusions to LDAP dependency to prevent classpath conflicts

saschaszott commented on dspace-api/src/main/java/org/dspace/authenticate/LDAPAuthentication.java:
```diff
@@ -446,252 +441,129 @@ private static class SpeakerToLDAP {
+            try {
+                boolean authenticated = ldapTemplate.authenticate(
+                    LdapUtils.newLdapName(dn),
+                    new EqualsFilter(ldap_id_field, "*").toString(),
```
If this is about checking whether an attribute exists in the LDAP entry, then it would be better to use a `PresentFilter`.

JohnnyMendesC replied:
Agreed. Switched to `PresentFilter` for checking attribute existence.

saschaszott commented on dspace-api/src/main/java/org/dspace/authenticate/LDAPAuthentication.java:
```diff
@@ -446,252 +441,129 @@ private static class SpeakerToLDAP {
+            this.ldapTemplate.setIgnorePartialResultException(true);
+        }
 
+        protected String getDNOfUser(String adminUser, String adminPassword, Context context, String netid) {
```
Method parameters `adminUser` and `adminPassword` are obsolete.

JohnnyMendesC replied:
Fixed. Removed the obsolete parameters (`adminUser`, `adminPassword`) from the method signature. It now relies on the `SpeakerToLDAP` instance configuration.

saschaszott commented on dspace-api/src/main/java/org/dspace/authenticate/LDAPAuthentication.java:
```diff
@@ -446,252 +441,129 @@
 
+            try {
+                boolean authenticated = ldapTemplate.authenticate(
+                    LdapUtils.newLdapName(dn),
```
In the calling method `authenticate`, the `dn` is in one case directly constructed from user input using `idField + "=" + netid + "," + objectContext`. No escaping takes place there. Either `netid` is escaped at that point and an `EqualsFilter` with `idField` is used, or this handling must be done here to ensure that LDAP query injection is not possible.

JohnnyMendesC committed:
Fix: Address code review (DN escaping, PresentFilter, unused methods)

JohnnyMendesC commented on dspace-api/src/main/java/org/dspace/authenticate/LDAPAuthentication.java:
```diff
@@ -446,252 +441,129 @@ private static class SpeakerToLDAP {
-                    // Use anonymous authentication
-                    env.put(javax.naming.Context.SECURITY_AUTHENTICATION, "none");
-                }
+                EqualsFilter filter = new EqualsFilter(ldap_id_field, netid);
```
I investigated the `EqualsFilter` source code, and I found that internally it calls `LdapEncoder.filterEncode()`. 

_It calls the super on `CompareFilter`:_
<img width="797" height="366" alt="image" src="https://github.com/user-attachments/assets/411e797b-2190-479c-80f1-cb0bb6765362" />

_And then encodes it with the `filterEncode`_
<img width="802" height="706" alt="image" src="https://github.com/user-attachments/assets/945c53e7-0861-4e3c-9e1d-e6146e516a19" />

I understood that adding it here explicitly would cause double-escaping and break the search, so I relied on `EqualsFilter` handling it.

JohnnyMendesC commented on dspace-api/src/main/java/org/dspace/authenticate/LDAPAuthentication.java:
```diff
@@ -446,252 +441,129 @@ private static class SpeakerToLDAP {
+                if (authenticated) {
+                    return true;
+                }
+                // Fallback para tentativa direta (útil em alguns cenários de AD)
```
Fixed. Translated all code comments to English.

JohnnyMendesC committed:
Fix: Address code review (throw exception on missing URL)

JohnnyMendesC committed:
Fix: Upgrade Spring LDAP to 3.3.5 and disable LDAP health check in tests

- Upgraded  to 3.3.5 to resolve classpath conflicts with Spring Boot 3.5.x.
- Removed dependency exclusions in  to restore  required by Actuator auto-configuration.
- Added explicit  dependencies to parent POM to resolve convergence errors.
- Disabled  in  to prevent HealthIndicatorsIT failure (DOWN status) due to missing LDAP server in test environment.

tdonohue reviewed (commented):
@JohnnyMendesC : I gave this code a quick look today and it looks good to me. I only have one very minor suggestion inline below.

However, before merging this, I'd want to verify that either @saschaszott or you have tested this with an LDAP instance.  I don't have an LDAP instance to test with, but I feel we *must* test this fully to ensure there's no change in behavior in LDAP Authentication.

Once it's verified to be tested with LDAP & the minor changes below are made, then I'd be +1 as well.

JohnnyMendesC committed:
Docs: Add comment explaining micrometer dependency usage

JohnnyMendesC commented:
@tdonohue Thanks for the final review!

1.  **Code Update:** I've added the explanatory comment to the `pom.xml` regarding the Micrometer dependency as requested.

2. **Regarding Manual Testing:** I currently don't have access to a live LDAP environment. I attempted to set up a local Docker-based OpenLDAP instance for verification, but unfortunately, I encountered infrastructure configuration issues unrelated to the code changes that prevented me from successfully completing the login flow manually (both on branch main and on the fix).

I fully agree that validating this against a real instance is important to ensure no behavioral changes. Since I am currently relying on the Integration Test suite, I would appreciate any help in validating this, or guidance on how to proceed if further verification is required before the merge.

tdonohue commented:
@alanorth : That'd be great if you can help test it!  From reviewing the code here, I think we really just need a basic test that LDAP Authentication is working, and that LDAP fields (name, etc) and groups are being synced over to DSpace.  There's no other special behaviors we need to test.

alanorth commented:
Dear @JohnnyMendesC and @tdonohue. I've tested this with our LDAP server and login functionality is working:

```
2026-02-10 09:46:53,166 INFO  unknown unknown org.dspace.authenticate.LDAPAuthentication @ me@example.com::authenticate:type=ldap
2026-02-10 09:46:53,166 INFO  unknown unknown org.dspace.eperson.EPersonServiceImpl @ me@example.com::update_eperson:eperson_id=a8df75ef-11d2-4875-afad-6ca3d7e49ecd
2026-02-10 09:46:53,167 INFO  unknown unknown org.dspace.app.rest.security.EPersonRestAuthenticationProvider @ me@example.com::login:type=explicit
```

The functionality before and after the patch is the same as far as I can tell.

P.S. I've fixed the merge conflicts in `pom.xml` related to the recent Java 17 → Java 21 and Spring Boot updates.

alanorth reviewed (approved):
Approved by testing LDAP login

tdonohue commented:
This couldn't be automatically backported.  @JohnnyMendesC, if you have time/interest, I think it'd be good to backport this to the `dspace-9_x` branch _at a minimum_.  Let me know if that's something you can help with or not.

(I also would welcome anyone to backport it to 8.x/7.x as well, if anyone has interest.  Per our [support policy](https://wiki.lyrasis.org/display/DSPACE/DSpace+Software+Support+Policy), though, we don't guarantee backports to those older releases.

JohnnyMendesC commented:
Thanks @tdonohue, @saschaszott  and @alanorth for all your contributions, testing, and guidance on this PR!

@tdonohue I'd be happy to help with the backport to dspace-9_x. 
Since this is my first time performing a manual backport, I just want to double-check the workflow: 

I should create a new branch based on dspace-9_x, cherry-pick the commits (resolving the conflicts mentioned by the dspace-bot), push to my fork, and then open a new Pull Request targeting the dspace-9_x branch. Is that correct?

I'll start working on it right away.

JohnnyMendesC commented:
@tdonohue just an update on the backports:

`dspace-8_x`: I have successfully opened the backport PR (https://github.com/DSpace/DSpace/pull/11895). The cherry-pick proccess was almost identical to the 9x one, with conflicts due to the `main` having older versions of Spring dependencies than `9_x` and `8_x` in the `LICENSES_THIRD_PARTY` file.
* `main`
https://github.com/DSpace/DSpace/blob/bc1cc23cd6e32cabf4929421e82a0605295fb8ce/LICENSES_THIRD_PARTY#L454
* `9_x`
https://github.com/DSpace/DSpace/blob/2e94982d35e76c5b436fc78b35a90883625db07d/LICENSES_THIRD_PARTY#L455
* `8_x`
https://github.com/DSpace/DSpace/blob/668b0edd89d2fb7d5da79a7d80c4f95d5578d780/LICENSES_THIRD_PARTY#L408
---
`dspace-7_x`: I won't have the availability to tackle this one at the moment. So, a note for future contributors: 
The `7_x` branch runs on an older Spring version `5.3.39` and `java 11`, unlike the `8_x/9_x/main` ones that run on `6.2.15` and `java 17/21`, so a direct cherry-pick will likely fail.
* `7_x`
https://github.com/DSpace/DSpace/blob/2039318e22a894caf2c71cc8dc873706b91085e6/pom.xml#L20-L23
* `8_x`
https://github.com/DSpace/DSpace/blob/668b0edd89d2fb7d5da79a7d80c4f95d5578d780/pom.xml#L20-L23
* `9_x`
https://github.com/DSpace/DSpace/blob/2e94982d35e76c5b436fc78b35a90883625db07d/pom.xml#L20-L24
* `main`
https://github.com/DSpace/DSpace/blob/bc1cc23cd6e32cabf4929421e82a0605295fb8ce/pom.xml#L20-L24
This means it will probably require downgrading the spring-ldap dependency to a version older than `LDAP 3.x` (since it requires Spring `6.x` with `java 17`) like the `LDAP 2.4.x` (that requires Spring `5.3.x` with `java 8`) that is compatible with `dspace7_x`. But that means it will likely require to adjust the refactored code (`LDAPAuthentication.java`) to handle any API differences between Spring LDAP 2.x and 3.x.
